src/OpenSage.Game/Logic/Player.cs

```diff
@@ -49,21 +54,109 @@ public void SelectUnits(IEnumerable<GameObject> units, bool additive = false)
             }
         }
 
+        private static Player FromMapData(Data.Map.Player mapPlayer, ContentManager content)
+        {
+            var side = mapPlayer.Properties["playerFaction"].Value as string;
+
+            // We need the template for default values
+            var template = content.IniDataContext.PlayerTemplates.Find(t => t.Side == side);
+
+            var name = mapPlayer.Properties["playerName"].Value as string;
+            var displayName = mapPlayer.Properties["playerDisplayName"].Value as string;
+            var translatedDisplayName = content.TranslationManager.Lookup(displayName);
+
+            var isHuman = (bool) mapPlayer.Properties["playerIsHuman"].Value;
+
+            var colorRgb = mapPlayer.Properties.GetPropOrNull("playerColor")?.Value as uint?;
+
+            ColorRgb color;
+
+            if (colorRgb != null)
+            {
+                color = ColorRgb.FromUInt32(colorRgb.Value);
+            }
+            else if (template != null) // Template is null for the neutral faction
+            {
+                color = template.PreferredColor.ToColorRgb();
+            }
+            else
+            {
+                color = new ColorRgb(0, 0, 0);
+            }
+
+            return new Player
+            {
+                Side = side,
+                Name = name,
+                DisplayName = translatedDisplayName,
+                Color = color,
+                IsHuman = isHuman
+            };
+        }
+
+        // This needs to operate on the entire player list, because players have references to each other
+        // (allies and enemies).
+        public static IEnumerable<Player> FromMapData(Data.Map.Player[] mapPlayers, ContentManager content)
+        {
+            var players = new Dictionary<string, Player>();
+            var allies = new Dictionary<string, string[]>();
+            var enemies = new Dictionary<string, string[]>();
+
+            foreach (var mapPlayer in mapPlayers)
+            {
+                var player = FromMapData(mapPlayer, content);
+                players[player.Name] = player;
+                allies[player.Name] =
+                    (mapPlayer.Properties.GetPropOrNull("playerAllies")?.Value as string)?.Split(' ');
+                enemies[player.Name] =
+                    (mapPlayer.Properties.GetPropOrNull("playerEnemies")?.Value as string)?.Split(' ');
+            }
+
+            foreach (var (name, player) in players)
```

Comment:
Nice, I didn't realise this tuple syntax worked for `KeyValuePair<T>`.

Reply:
It works because of `OpenSage.Utilities.Extensions.KeyValuePairExtensions`, but it'd work out-of-the-box with .NET Core.